sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -280,28 +316,77 @@ public void processElement(ProcessContext processContext) throws Exception {
       while (!finished) {
         ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
         List<String> keys = scanResult.getResult();
-
-        Pipeline pipeline = jedis.pipelined();
-        if (keys != null) {
-          for (String key : keys) {
-            pipeline.get(key);
-          }
-          List<Object> values = pipeline.syncAndReturnAll();
-          for (int i = 0; i < values.size(); i++) {
-            processContext.output(KV.of(keys.get(i), (String) values.get(i)));
-          }
+        for (String k : keys) {
+          processContext.output(k);
         }
-
         cursor = scanResult.getStringCursor();
         if ("0".equals(cursor)) {
           finished = true;
         }
       }
     }
+  }
+  /** A {@link DoFn} requesting Redis server to get key/value pairs. */
+  private static class ReadFn extends BaseReadFn<KV<String, String>> {
+    @Nullable transient Multimap<BoundedWindow, String> bundles = null;
+    @Nullable AtomicInteger batchCount = null;
+    private final int batchSize;
+
+    @StartBundle
+    public void startBundle(StartBundleContext context) {
+      bundles = ArrayListMultimap.create();
+      batchCount = new AtomicInteger();
+    }
 
-    @Teardown
-    public void teardown() {
-      jedis.close();
+    ReadFn(RedisConnectionConfiguration connectionConfiguration, int batchSize) {
+      super(connectionConfiguration);
+      this.batchSize = batchSize;
+    }
+
+    private int getBatchSize() {
+      return batchSize;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext processContext, BoundedWindow window)
+        throws Exception {
+      String key = processContext.element();
+      bundles.put(window, key);
+      if (batchCount.incrementAndGet() > getBatchSize()) {
+        Multimap<BoundedWindow, KV<String, String>> kvs = fetchAndFlush();
```

Comment:
why you need to deal with windows here ? (note I quickly looked but didn't get the intuition), if we can avoid this probably is better, no?

Reply:
the window stored in the key here is used in FinishBundle to output the keys, since the the context in FinishBundle takes window as a parameter. `context.output(kv, w.maxTimestamp(), w);`